File: Proxy.cpp
/*****
 * Proxy
 * Provides a proxy that allows other objects to control access to this object.
*****/

#include <iostream>
#include <memory>

using namespace std;


class Subject { public: virtual void method() = 0; };

class ConcreteSubject : public Subject { public: void method() override { cout << "..." << endl; } };

class Proxy : public Subject {
    unique_ptr<ConcreteSubject> concreteSubject;

    bool validateAccess() { return true; }

public:
    Proxy() : concreteSubject(make_unique<ConcreteSubject>()) {}

    void method() override {
        if (validateAccess()) {
            cout << "Proxy: Start" << endl;
            concreteSubject->method();
            cout << "Proxy: End" << endl;
        }
    }
};


int main() {
    auto client = make_unique<Proxy>();
    client->method();
    
    return 0;
}

File: Proxy.java
/*****
 * Proxy
 * Provides a proxy that allows other objects to control access to this object.
*****/

interface Subject { void method(); }

class ConcreteSubject implements Subject {
    @Override
    public void method() { System.out.println("..."); }
}

class Proxy_ implements Subject {
    private ConcreteSubject concreteSubject;

    public Proxy_() { this.concreteSubject = new ConcreteSubject(); }

    private boolean validateAccess() { return true; }

    @Override
    public void method() {
        if (validateAccess()) {
            System.out.println("Proxy: Start");
            concreteSubject.method();
            System.out.println("Proxy: End");
        }
    }
}


public class Proxy {
    public static void main(String[] args) {
        Subject client = new Proxy_();
        client.method();
    }
}
